[release/3.1] Fix libunwind linking with gcc10, clang11 or -fno-common

omajid opened the pull request:
Linking coreclr with gcc 10 or clang 11 (both of which default to -fno-common) fails with this error:

```
[31%] Linking CXX shared library libdbgshim.so                                                                                                               
/usr/bin/ld: ../../pal/src/libcoreclrpal.a(dyn-info-list.c.o):/home/omajid/devel/dotnet/coreclr/src/pal/src/libunwind/src/mi/dyn-info-list.c:28: multiple definition of `_U_dyn_info_list'; ../../pal/src/libcoreclrpal.a(Linit.c.o):/home/omajid/devel/dotnet/coreclr/src/pal/src/libunwind/src/x86_64/Ginit.c:52: first defined here
```

This PR fixes that.

~~This is of https://github.com/libunwind/libunwind/pull/166, with one change: there's no s390x file in coreclr's copy of libunwind, so I removed that chunk. Otherwise the change applies as-is.~~

~~This was also added to runtime 5.0 with https://github.com/dotnet/runtime/commit/581dc193406082f76ae3c58511f724ab5d5c0d6d#diff-c144efc97b24537a136fc6842bad278829924e76dffa2f09ad1d8f078b8c116b, as part of the larger update to libunwind 1.5~~

# Customer impact
Customers like Redhat cannot build .NET Core using clang11 without this fix.

# Regression?
No

# Testing
Redhat side testing.

# Risk
Low, the change just adds a compiler option that doesn't affect the behavior on Clang older than 11 (the ones that we use for our product builds)

janvorli commented:
@omajid I have thought I have already commented on this PR, but I have probably forgotten to push the comment button. Are you still modifying it or have you just pushed a dummy change to restart the CI?

omajid commented:
I just pushed a dummy change (actually, updated the commit message to be more detailed) to restart CI.

janvorli commented:
@omajid can you please provide more details on why you need this change in 3.1?

omajid commented:
> @omajid can you please provide more details on why you need this change in 3.1?

Sure.

We are trying to build and distribute .NET Core 3.1 with Fedora. The current in-development version of Fedora (34) includes Clang 11.1.0 RC 2. This version fails to compile this piece of code:  

```
/usr/bin/ld: ../../pal/src/libcoreclrpal.a(dyn-info-list.c.o):/coreclr/src/pal/src/libunwind/src/mi/dyn-info-list.c:28: multiple definition of `_U_dyn_info_list'; ../../pal/src/libcoreclrpal.a(Linit.c.o):/coreclr/src/pal/src/libunwind/src/x86_64/Ginit.c:52: first defined here                              
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)                                                                       
make[2]: *** [src/dlls/dbgshim/CMakeFiles/dbgshim.dir/build.make:128: src/dlls/dbgshim/libdbgshim.so] Error 1                                            
make[1]: *** [CMakeFiles/Makefile2:57639: src/dlls/dbgshim/CMakeFiles/dbgshim.dir/all] Error 2                                                           
make[1]: *** Waiting for unfinished jobs....    
```

To reproduce:

```bash
docker run -it registry.fedoraproject.org/fedora:rawhide /bin/bash
sudo dnf install -y clang cmake git krb5-devel libcurl-devel lldb-devel libicu-devel llvm make zlib-devel lttng-ust-devel userspace-rcu-devel zip unzip wget curl python36 tar openssl-devel bash-completion nodejs tmux libunwind-devel findutils
git clone https://github.com/dotnet/coreclr -b release/3.1
cd coreclr
./build.sh
```


The [Clang 11 release notes](https://releases.llvm.org/11.0.0/tools/clang/docs/ReleaseNotes.html#modified-compiler-flags) say that `-fno-common` is now the default, just like it is for GCC 10.

> -fno-common has been enabled as the default for all targets. Therefore, C code that uses tentative definitions as definitions of a variable in multiple translation units will trigger multiple-definition linker errors. Generally, this occurs when the use of the extern keyword is neglected in the declaration of a variable in a header file. In some cases, no specific translation unit provides a definition of the variable. The previous behavior can be restored by specifying -fcommon.

So this fix (also present in 5.0 and later) takes care of this.

We have been carrying this as a local patch in Fedora for a few months now: https://src.fedoraproject.org/rpms/dotnet3.1/c/fac26bd1ab1af64a08f3b9076bcf5d4505842947?branch=master

Alternatively, we could set `-fcommon`, but I dont know the impact of that throughout the codebase.

janvorli commented:
It seems that `-fcommon` was the default for previous versions of clang, so it should be no problem to use that option explicitly. It would be preferable to making the change to 3.1. We have quite a high bar for changes going to 3.1.

aik-jahoda commented:
@omajid, @janvorli can we push this PR further? It seems we need servising and reviewer approval.

omajid commented:
I need to re-do this PR entirely based on @janvorli 's insightful feedback. I have been putting it off for a bit but I will try getting to this again.

omajid committed:
Fix compilation under newer GCC/Clang

Linking coreclr with gcc 10 or clang 11 (both of which default to
-fno-common) fails with this error:

    [31%] Linking CXX shared library libdbgshim.so
    /usr/bin/ld: ../../pal/src/libcoreclrpal.a(dyn-info-list.c.o):src/pal/src/libunwind/src/mi/dyn-info-list.c:28:
    multiple definition of `_U_dyn_info_list'; ../../pal/src/libcoreclrpal.a(Linit.c.o):src/pal/src/libunwind/src/x86_64/Ginit.c:52:
    first defined here

This commit fixes that, by switching to -fcommon as the default.

omajid commented:
@janvorli WDYT?

janvorli commented:
@omajid I am sorry for the delay, I was on a paternity leave for the last month.

janvorli reviewed (approved):
LGTM, thank you!

jeffschwMSFT reviewed (approved):
Approved.  We will take for consideration in 3.1.x

omajid commented:
@janvorli Congratulations! No worries about a delay; this wasn't blocking anything.
